Implementation of a `loc` keyword argument (similar to that of `matplotlib`'s label setters) for the `wcsaxes` `AxisLabels`

athob opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to propose the implementation of a `loc` keyword argument to control label position in a `WCSAxes` similar to that available in `matplotlib`'s label setters.

In `matplotlib`, the label setters allow the assignment of a `loc` keyword argument as a high-level alternative for passing class `Text` parameters `horizontalalignment` (determining the `Text` object horizontal alignment), and `x` or `y` (determining the `Text` object position) depending if the setter affect the label of the x or y axis. The value that are allowed are either [`'left'`, `'center'` or `'right'` for the xlabel setter](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_xlabel.html), or [`'bottom'`, `'center'` or `'top'` for the ylabel setter](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_ylabel.html), each intuitively providing control as to which position the label will take relative the corresponding axis.

On the other hand, while the `astropy.visualization.wcsaxes.core.WCSAxes` provide equivalent label setting methods via the method `set_axislabel` of the its `astropy.visualization.wcsaxes.coordinate_helpers.CoordinateHelper` API, neither this method nor the class' other methods provide a satisfyingly equivalent customization mean. It is technically possible to manually set the `x`, `y` and `horizontalalignment` [`matplotlib.text.Text`](https://matplotlib.org/stable/api/text_api.html#matplotlib.text.Text) attributes of the underlying corresponding label class `astropy.visualization.wcsaxes.axislabels.AxisLabels` (an inherited class of `Text`) from the `wcsaxes` API, but in the current implementation, `x` and `y` eventually get overwritten anyway by the class `draw` method, forcing the label to be centered along the axis length.

This implementation add the `loc` keyword argument to the definition of class `AxisLabels`, with its assignment accessible to the user via the method `CoordinateHelper.set_axislabel` which, as mentioned, is currently already the method of use to assign both the label text and padding for that class. The goal of the implementation is to be as equal as possible to that of `matplotlib`, while retaining a similar style of customization to that implemented for the `minpad` attribute (where a dictionary of different parameters can be given to set different value for each visible axis). In order to be fully functional and satisfyingly modular enough, some choices needed to be made, some of which affecting past functionalities that were most likely niche and probably already inconsistent:
- the past implementation allowed `horizontalalignment` (or `ha`) to be set by the user, which gave little to no control, as the most important attributes to adequately position the label were the position ones `x` and `y` that are overwritten in the method `draw`; the newer implementation leaves its unrestricted control to the `loc` definition.
- the past implementation was inconsistent with the `matplotlib`'s convention to have the `verticalalignment` set by default to `baseline` for all labels; the new implementation follow now that convention (which also caused the bbox_tests to accordingly be modified as the old reference extents were incompatible)
- the past implementation was also partially inconsistent with the `matplotlib`s convention to set the `rotation_mode` to `'anchor'` in the case of y-axis labels with vertical rotation; in order to be fully compatible with the multi-axes label drawing feature of the `AxisLabels` API, the `rotation_mode` is set by default to `'anchor'` for all axes (note that the user may still change if if he wishes, even if if it may lead to unexpected results and is likely unadvisable)
- the newer implementation includes a warning printed to the user in the event that they attempt to set in `CoordinateHelper.set_axislabel` any of the Text low-level attributes that are controlled by `loc` (those include `x`, `y`, `rotation`, `horizontalalignment` & `ha`, and `rotation_mode` - the latter to disincentivize its usage)

Allowing this implementation to be fully functional required the method `_halfway_x_y_angle` of class `astropy.visualization.wcsaxes.frame.Spine` to be repurposed and generalized into a new method `_barycentric_x_y_angle`. The method `_halfway_x_y_angle` was previously used to return the x, y, and rotation angle corresponding to the very center of an axis Spine object, which was then used to center the axis label along that axis at the drawing stage. The method `_barycentric_x_y_angle` does the same thing but using instead a given scale parameter between 0 and 1, with each value corresponding to either end of the axis Spine (making `_halfway_x_y_angle` a special case at scale factor equal to 0.5). Its implementation was also adapted to always return that 0 to 1 position assuming a general orientation of the Spine from the lower left corner of the figure to its upper right corner (except if the Spine vertices do not follow a generally defined direction, such as in the case of an elliptical spine). For the purpose of the `loc` keyword argument implementation, the `_barycentric_x_y_angle` method is replacing `_halfway_x_y_angle` with an input determined by the `loc` value.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

<!-- Fixes #<Issue Number> -->

<!-- Optional opt-out -->

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

athob committed:
State of PR when submitted (rebased squash of multiple commits)

Small assignment of halfway distance in _halfway_x_y_angle

Generalizing method _halfway_x_y_angle to a newer method returning the corresponding output at any barycentric coordinate position along the spine

Small consistency coding update

Preparing the AxisLabels class definition to implement a new loc kwarg similar to that of matplotlib label setters

Updating further the new loc kwarg implementation, allowing it to depend on the axislabel axis, and moving the setting of horizontalalignment and verticalalignment to the drawing method - also adding a warning regarding the overwriting of loc's low level kwargs in case those are set by the user

Forcing axislabels to have anchor rotation mode to be consistent with pyplot axes labels, notably for y-axes, and allowing modification of verticalalignment, adapting the protect_kw list as well

Reposition lines of code

Fixing _barycentric_x_y_angle to consider orientation of the spine - so that bary 0 or 1 respectively correspond to lower left and upper right corner

Small refactor

Reposition line of code

Full implementation of the new loc AxisLabels kwarg

Documenting the new loc kwarg

Fixing ruff flagged formatting

Adding a changelog entry for PR #18749

neutrinoceros commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -1,16 +1,17 @@
 
 class AxisLabels(Text):
-    def __init__(self, frame, minpad=1, *args, **kwargs):
+    def __init__(self, frame, minpad=1, loc="center", *args, **kwargs):
```
This argument should be keyword only. If only to mimic matplotlib's API
```suggestion
    def __init__(self, frame, minpad=1, *args,  loc="center", **kwargs):
```
It'll also avoid breaking backward compatibility with anyone passing positional args currently captured in `*args`

neutrinoceros commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -23,17 +24,23 @@ def __init__(self, frame, minpad=1, *args, **kwargs):
+        self.set_minpad(minpad)
+        self.set_loc(loc)
+        self.set_rotation_mode("anchor")
+        self.set_visibility_rule("labels")
```
please check this thoroughly, but I'd expect this could be wrapped as a single method call as follows

```suggestion
        self.set(
            minpad=minpad,
            loc=loc,
            rotation_mode="anchor",
            visibility_rule="labels",
        )
```

neutrinoceros commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -90,15 +100,45 @@ def draw(
             # along the path but in future we could allow this to be a
             # parameter.
-            x, y, normal_angle = self._frame[axis]._halfway_x_y_angle()
+            x, y, normal_angle = self._frame[axis]._barycentric_x_y_angle(bary)
```
this looks orthogonal to what you're trying to achieve in this PR, and I suspect this could be what's breaking existing image tests. Can you explain (or possibly, revert ?) ?

neutrinoceros commented on astropy/visualization/wcsaxes/coordinate_helpers.py:
```diff
@@ -716,7 +716,7 @@ def set_ticklabel_visible(self, visible):
         self._ticklabels.set_visible(visible)
 
-    def set_axislabel(self, text, minpad=1, **kwargs):
+    def set_axislabel(self, text, minpad=1, loc="center", **kwargs):
```
here too, this should never be allowed positionally

```suggestion
    def set_axislabel(self, text, minpad=1, *, loc="center", **kwargs):
```

neutrinoceros commented on astropy/visualization/wcsaxes/coordinate_helpers.py:
```diff
@@ -739,10 +743,27 @@ def set_axislabel(self, text, minpad=1, **kwargs):
+            "ha",
+            "rotation_mode",
+        ]
+        if {*kwargs} & {*protected_kw}:
```
while I do find your original implementation very elegant, using plain english would probably help more casual readers understanding what's happening here 
```suggestion
        if set(kwargs).intersection(protected_kw):
```

neutrinoceros commented on astropy/visualization/wcsaxes/coordinate_helpers.py:
```diff
@@ -739,10 +743,27 @@ def set_axislabel(self, text, minpad=1, **kwargs):
+            "rotation_mode",
+        ]
+        if {*kwargs} & {*protected_kw}:
+            warnings.warn(
```
I wonder if this should be an exception instead. If a warning is indeed prefered, it should use `category=AstropyUserWarning` (from `astropy.utils.exceptions`) and set `stack_level` explicitly.

athob replied:
Updated in commit 65ea4c6

neutrinoceros commented on astropy/visualization/wcsaxes/frame.py:
```diff
@@ -114,21 +114,43 @@ def _halfway_x_y_angle(self):
         """
+        return self._barycentric_x_y_angle(0.5)
+
+    def _barycentric_x_y_angle(self, bary=0.5):
```
```suggestion
    def _barycentric_x_y_angle(self, bary=0.5, /):
```

neutrinoceros replied:
it would also be a nice opportunity to add type annotations to this signature

athob replied:
Thanks, I'll leave it without additional annotations then for now, and focus on other tasks. I'll consider making another PR for the annotations in the future

neutrinoceros reviewed (changes requested):
thank you very much for working on this. I haven't identified the cause of the regression yet but here's a first pass of comments, questions and suggestion while I get familiar with this.

athob committed:
Addressing first set of code review (rebased squash of multiple commits)

Narrowing down the values allowed for `axis` when processing the `loc` parameter

Co-authored-by: Clément Robert <cr52@protonmail.com>

Same as previous commit e4dce48 for a different `axis` case

Co-authored-by: Clément Robert <cr52@protonmail.com>

Fixing reStructuredText formatting [ci skip]

Co-authored-by: Clément Robert <cr52@protonmail.com>

Updating argument and keyword argument positioning in function signatures

Co-authored-by: Clément Robert <cr52@protonmail.com>

Apply various implementation improvements from the review suggestions

Co-authored-by: Clément Robert <cr52@protonmail.com>

Make use of the existing `set` method of (grand)parent class `Artist`

The implementation of `Artist.set` is compatible with the assignment of `AxisLabels` exclusive attributes `minpad`, `loc` and `visibility_rule`

Co-authored-by: Clément Robert <cr52@protonmail.com>

Annotating the signatures of _barycentric_x_y_angle and _halfway_x_y_angle

Fixing typo & revising formatting

Generalizing the proposed use of the existing `set` method of `AxisLabels`

Adding proper category for warning issued on low level  kwargs

Quick ruff formatting fix

athob commented:
> @athob note that it is possible to batch several suggestions into a single commit on GitHub's web UI, but only from the "files" tab of the PR page

Thanks @neutrinoceros, there's a lot I'm unfamiliar with in the reviewing interface. I'll comment on everything as soon as I've handled the suggestions that don't require much discussion.

athob commented:
> @athob - would you be able to check the output of the figure for `test_bbox_size` for before and after the change to see if the change makes sense? (if you could post the output figures here that would be great - you can save figures by putting e.g. `fig.canvas.savefig('test.png')` inside the test.

Thanks @astrofrog, I had checked visually the change to the rendered figures before modifying the test and those eventually slightly shifted the labels vertically (relative to their baseline, i.e. vertically for x-label and horizontally for y-label relative to the canvas), affecting the overall canvas bbox extent. At the end (see below), I am going to revert the test to its original implementation as I made a mistake while assessing the need to change the `verticalalignment` default.

> Regarding the discussion on the default behavior, I believe it comes down to the default setting of the 2 parameters `verticalalignment` and `rotation_mode` of `AxisLabels`. I'll describe a bit more in details the reasoning that motivated me in changing them (although I'm revising my opinion regarding `verticalalignment` after reviewing a bit more the way `matplotlib`'s label setters are working).

@neutrinoceros I'm following up regarding the default behavior discussion, I was double checking a couple things regarding my choices and what brought me to make them. As discussed, the default change only concern the default assigned value to the 2 `AxisLabels` parameters:
- `verticalalignment` used to be set by default to `'center'` and I chose to unset it (leaving it to its `Text` default value `'baseline'`)
- `rotation_mode` used to be left unset (leaving it to its `Text` default value `'default'`), and I chose to set it by default to `'anchor'`
Those choices were motivated by wanting to replicate the `matplotlib` label setters implementation, and in order to facilitate the development of the new `loc` kwarg.

However after reviewing these 2 parameters, I'm now going back on my decision to change the `verticalalignment` but I keep believing changing the `rotation_mode` default would be beneficial, in particular in the context of this new feature.

### - regarding `verticalalignment` - it should be kept by default to `'center'` (the current behavior)

When starting planning on how to implement the `loc` parameter, I looked into the `matplotlib` source code to figure how they do it. I reviewed the way the `loc` parameter intervene after being passed to the [`set_xlabel` method](https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axes/_base.py#L3596-L3645) and the [`set_ylabel` method](https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axes/_base.py#L3849-L3898) into configuring the corresponding label `Text` objects (you may recognize some of the implementation choices I have made as those were inspired by `matplotlib`'s).

Because neither setter seem to touch the `verticalalignment` `Text` attribute, I (prematurely) assumed those remained unset and would take the default value `'baseline'` set by the `Text` class, and then I continued development with this assumption. Had I reviewed how the corresponding label `Text` objects get first configured by default [at the construction of the XAxis object](https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axis.py#L2369-L2406), and [at the construction of the YAxis object](https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axis.py#L2593-L2632), I would have noticed that the `verticalalignment` does get actually set with a different value than the default class `'baseline'`: specifically, `matplotlib` set it to `'top'` for the `XAxis` and `'bottom'` for the `YAxis`.

Now, it is "easy" for `matplotlib` to use distinct default values because of the distinct purposes of both the `XAxis` and `YAxis`. When it comes to `wcsaxes`' `AxisLabels`, it's not as trivial because `AxisLabels` must represent labels of all axes in any orientation. The `verticalalignment` should accordingly be defaulted to a unique value that would work for any of those orientation, and after that review, `'center'` seems to fill that purpose better than the default `Text` `'baseline'`. Sorry that I overlooked this.

### - regarding `rotation_mode` - it should be changed by default to `'anchor'` (my proposed behavior)

As I was continuing development of the `loc` feature, I was struggling to properly control how the alignments where coming into play after the `draw` method was setting the `rotation`, `x` and `y` attribute of the `AxisLabels`. Replicating the way the `matplotlib` setters were setting the `horizontalalignment` attribute did not produce the intended result (mostly with y-axis labels).

After debugging and testing, I noticed that one of the cause of that struggle was the fact that the `rotation_mode` was left unset to `'default'`, and not set to `'anchor'`. After doing that change, it made the `horizontalalignment` more intuitive in the way it controlled the label layout. The `'anchor'` value is also the default value used in the [`YAxis`](https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axis.py#L2593-L2632) constructor. Note that the  [`XAxis`](https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axis.py#L2369-L2406) constructor does not set it, but it also does not make any rotation.

I have done a few tests since yesterday locally, and setting the `rotation_mode` attribute to `anchor` passes the bbox test. It may however break at with the CircleCI image tests. I want to see if it does, so I will work on an update that will revert the default behavior for the `verticalalignment` and keep the `'anchor'` default for the `rotation_mode` attribute. I'll push it before end of day.

athob committed:
Improving the new kwarg `loc` feature (rebased squash of multiple commits)

Reverting to default verticalalignment value set to

Updating user warning's stacklevel to 2

Revising `loc` support for `'c'` axis
Future updates should reconsider the purpose of `loc` in the case of an `EllipticalFrame` after the label drawing is fixed for axis `'c'`

Refining the documentation of `set_axislabel` to detail the dictionary configurability of its visual property parameters

Replacing use of try/except statements in visual property getters

Changing the default of the `loc` kwarg to `None`
This will set it to the corresponding entry in the `pyplot.rcParams` dictionary - which for a clean installation should be `'center'`

Updating the signature to not include `loc`
The `loc`` kwarg does not need to be explicitely included in the signature, and explicitely set via `set_loc``, since the `set` method inherited from `Artist` will if kwargs contains the `loc` entry. Explicitely having `loc` included and set via `set_loc` results in an unintended behavior that will overwrite `loc` if calling multiple time `set_axislabel` without specifying `loc`.

Fixing the `axis` and `loc` conditional statement in `draw`
Updating the NotImplementedError raising case for unsupported `loc` values, from flagging if `loc` is not `center`, to flagging if `loc` is not None

athob commented on astropy/visualization/wcsaxes/tests/test_misc.py:
```diff
@@ -578,7 +578,7 @@ def test_set_labels_with_coords(ignore_matplotlibrc, frame_class):
 def test_bbox_size(atol):
     # Test for the size of a WCSAxes bbox (only have Matplotlib >= 3.0 now)
-    extents = [11.38888888888889, 3.5, 576.0, 432.0]
+    extents = [7.38888888888889, 7.5, 576.0, 432.0]
```
This was reverted in commit fc4907b, I'll resolve for now

athob commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -23,17 +24,23 @@ def __init__(self, frame, minpad=1, *args, **kwargs):
+        try:
+            return self._loc[axis]
+        except TypeError:
+            return self._loc
```
I actually worked a bit on it, and got a newer version that I pushed in a "subbranch" for now for your review. The diff between this branch's state as base and that subbranch is available [here](https://github.com/athob/astropy/compare/implement-loc-kwarg-in-axislabels...athob:astropy:implement-loc-kwarg-in-axislabels.better_minpad_loc_getter).

I notably [improved the documentation of `set_axislabel`](https://github.com/athob/astropy/commit/86687cfe7c3aa93520e4b68d7f1c574d77e63589) to explicitely discuss the dictionary configurability feature, and modified the getters to [get rid of the `try/except` statements](https://github.com/athob/astropy/commit/f8e4f871a4590374e9cf4877da9a24b34a3ca9b3).

I also took the liberty to make a couple additional changes:
- I originally explicitly set the default `loc` to `'center'` but the default behavior should reflect `matplotlib` `rcParams`'s corresponding entries, which actually happens in `draw` when `loc` is `None`. Accordingly [I changed the `'center'` default of the kwargs to `None`](https://github.com/athob/astropy/commit/a3722e8e111b0a0443ca834d7ba5f5fc3282c9eb) (this should not affect the tests because a clean install of `matplotlib` will have `'center'` in the corresponding `rcParams` entries).
- [I upgraded the getters and setters](https://github.com/athob/astropy/commit/29952fba083b7d7e426b5ac7c9793a5ab70f4405) for the configuration dictionary case so that, on one hand, the getter can still return a value (the default) when the dictionary is missing an axis, and on the other hand, the setter raise a warning when given a dictionary that is incomplete relative to the visible axes. As a comment on that topic: I feel like the `minpad` default should perhaps be tied to the `rcParams['axes.labelpad']` in some ways, but that question is probably worth a PR of its own
- Lastly, [I realized that](https://github.com/athob/astropy/commit/804e5c5a67589c1b9a8e4a9a0208e3b7da5d6867) the signature and definition of `set_axislabel` do not need to explicitly list `loc` and call `set_loc` on it, as the `set` method will do if `kwargs` contains it. The `loc` is set by default by the `AxisLabels` constructor to `None`, and in the event that a user calls `set_axislabel` multiple time in a row (such as for modifying the `text`, but not change the `loc`), it will be frustrating if `loc` gets overwritten with `None` on the subsequent calls. On that note, the argument can also be made for any of the arguments of that function, `text` and `minpad` too, but those being positional makes it tricky from a backward compatibility perspective.

If the updates in this subbranch seem sound, I'll merge it into this PR branch.

athob replied:
> It's a bit hard to review a diff outside of a PR, where I don't have access to tools like suggestions and inline comments. Though, briefly
> 
> ```python
> if isinstance(self._minpad, dict):
>      return self._minpad.get(axis, 1)
> else:
>      return self._minpad
> ```
> 
> doesn't look equivalent to
> 
> ```python
> try:
>     return self._minpad[axis]
> except TypeError:
>     return self._minpad
> ```
> 
> because the original code might still raise a `KeyError` if the key is missing from `self._minpad`, while your proposed replacement silently returns a default value instead.

Got it, I'm going to rebase this subbranch's history to isolate the one commit (70f26f1) that switch from a key passed though subscription, to a key passed through the `get` method, and will merge the rest in the PR branch if it can help with the review. I think the `minpad` documentation still need to be more detailed, and the `try`/`except` statement can be avoided in the implementation without removing the `KeyError`.

I agree that it'd be better to have, at the very least, a warning sent to the user if the key isn't in the `minpad` dictionary rather than silently returning the default value. I tried to mitigate that by having the warning raised when using the `minpad` setter, based on the axes that are actually visible (returned by `get_visible_axes`), but those can be changed by the user between the use of the `minpad` setter and the drawing.

I'll think of something, but otherwise I'll leave that particular commit for another PR.

> > the default behavior should reflect matplotlib rcParams's corresponding entries
> 
> sounds like we need to query that default dynamically, so we don't have to keep it in sync manually !

In the current state of the PR branch, having `None` given to `loc` allow for the `rcParams` to be used, it happens per axis in the `draw` method (lines [107](https://github.com/athob/astropy/blob/8b31421076780feb5bfda9cc728ebacaa495e173/astropy/visualization/wcsaxes/axislabels.py#L107) and [116](https://github.com/athob/astropy/blob/8b31421076780feb5bfda9cc728ebacaa495e173/astropy/visualization/wcsaxes/axislabels.py#L116) in the `AxisLabels` class definition). So with the implementation of the `loc` getter for each axis, having `None` used as default assures that the `rcParams` entries get eventually used.

> > As a comment on that topic: I feel like the minpad default should perhaps be tied to the rcParams['axes.labelpad'] in some ways, but that question is probably worth a PR of its own
> 
> Agreed. Let's postpone this for now. We don't want this feature-creep this first PR.
> 
> > but those being positional makes it tricky from a backward compatibility perspective.
> 
> preaching to the choir here 😄

I'm giving myself a todolist of PRs!

athob commented:
> > I keep believing changing the rotation_mode default would be beneficial, in particular in the context of this new feature.
> 
> I think it makes sense too. Thanks for your thoughtful analysis !
> 
> I think the only missing bit for me to approve the PR now is a test to exercise the new feature.

Thanks, I'll try to design a test over the weekend. Is that something that could be integrated in the CircleCI image comparisons, and if it is, would you know where I should start to design those?

athob committed:
Designing a set of figure_tests for the new `loc` feature

athob commented:
Hi @neutrinoceros, I think I understood how to make new CircleCI image comparisons tests using the `figure_test` decorator. I designed 4 tests this way to test various configuration of the `loc` parameter. I wasn't sure how those would interact with the CircleCI checks, so for now I pushed the commit in a "subbranch" for your review. The comparison of that subbranch with this branch is available [here](https://github.com/athob/astropy/compare/implement-loc-kwarg-in-axislabels...athob:astropy:implement-loc-kwarg-in-axislabels.figure_tests), let me know if you have any question, or any suggestion or correction you'd like to see.

While designing the tests, I realized that the `"c"` axis for an `EllipticalFrame` didn't behave the way I thought it would, and I believe its label drawing may need an update (which would probably require another PR preceded by an issue discussion). Because of that, [I removed the `"c"` axis case from the `loc` check at drawing](https://github.com/astropy/astropy/pull/18749/commits/364365181cec3b251047c8af1a9d89bc10da396d). I may consider a more adequate set of supported `loc` values for this axis in the future. On a minor note, I also noticed that one of the `get_axislabel_visibility_rule` getter method of the `CoordinateHelper` class had a function signature requiring an argument, which seemed unusual for a getter, and that argument was not being used. As far as I can tell, this method isn't used anywhere, [so I took the liberty to remove the argument from the signature](https://github.com/astropy/astropy/pull/18749/commits/8b31421076780feb5bfda9cc728ebacaa495e173). I understand this could be a bit radical, I'm waiting to see if the tests are breaking because of this. If they break, or if you think removing this unused argument would be an issue, I can always revert that commit (this could also be a separate PR).

Lastly, [I just wanted to point out my previous discussion](https://github.com/astropy/astropy/pull/18749#discussion_r2461828421) regarding the implementation of a corrected/improved code for the `minpad` and `loc` setter/getter of `AxisLabels`. I rebased the corresponding subbranch to include the update I made to this branch. Let me know if the updates I described are acceptable additions as part of this PR, and if they are I'll merge the subbranch in this one.

athob commented:
Awesome, thanks for the review! I'll merge in the tests later today.

> > if you think removing this unused argument would be an issue
> 
> If there's even the slightest chance this (unused) argument could be passed somewhere in the wild, we should add a deprecation warning first. I don't think we have quite the appropriate decorator for this in `astropy.utils.decorators` but you can also craft the warning manually like so:
> 
> ```python
> import warnings
> from astropy.utils.exceptions import AstropyDeprecationWarning
> warnings.warn("...", category=AstropyDeprecationWarning, stacklevel=...)
> ```
> 
> > this could also be a separate PR
> 
> if you don't mind, that'd be great. We could almost immediately merge that one.

Alright, in that case, I'm gonna revert commit 8b31421 for now, and will create a separate PR including the warning after I'm done working on this one today. I also want to review the discussion on the improved `minpad`/`loc` getter/setter, will respond there.

athob commented:
Thanks @neutrinoceros for the review. I'll work on incorporating the changes as soon as I can, for now I'm gonna respond to some of the interrogations you had.

athob commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -90,15 +102,46 @@ def draw(
+            elif loc is not None:
+                raise NotImplementedError(
+                    f"Received unsupported value {loc=!r}. "
+                    f"Only loc='center' is implemented for {axis=!r}"
```
So before commit e973ba8, I used to have `'center'` be the default `loc` value at construction. In that commit, I changed that to `None` because it allows for a dynamic use of the default `loc` that is available in the `matplotlib` `rcParams` (in line [107](https://github.com/athob/astropy/blob/6374d7f008c6d503d19907c1dd6313a6561d0171/astropy/visualization/wcsaxes/axislabels.py#L107) and [116](https://github.com/athob/astropy/blob/6374d7f008c6d503d19907c1dd6313a6561d0171/astropy/visualization/wcsaxes/axislabels.py#L116)). But with the former behavior to have `'center'` as the default `loc`, since that `NotImplementedError` raising case used to check if `loc` was different from `'center'` and with now the default `loc` being `None`, it caused issues for any `axis` that does not fall under cases `axis in {"t", "b", "h"}` (horizontal axis) or `axis in {"l", "r", "v"}` (vertical axis) such as axis `"c"` (elliptical axis). I modified it then in commit 6374d7f so it would test `loc` against `None` rather than `'center'`, but kept the error message identical since, eventually, if it was `None`, it would be force to become `'center'` in [the catch-all case](https://github.com/athob/astropy/blob/6374d7f008c6d503d19907c1dd6313a6561d0171/astropy/visualization/wcsaxes/axislabels.py#L129-L131). But now that I think about it, there is a situation where this could break, in the situation where the `axis` does not fall in the `{"t", "b", "h"}` or `{"l", "r", "v"}`, and the user manually set `loc` as `'center'`. I'm gonna adapt it again so that the `elif loc is not None:` reads as `elif loc is not None and loc != 'center':` which should fix that specific case.

athob committed:
New code review suggestions (rebased squash of multiple commits)

Apply suggestions from code review

Co-authored-by: Clément Robert <cr52@protonmail.com>

Quick patch post-review

athob commented on astropy/visualization/wcsaxes/axislabels.py:
```diff
@@ -1,39 +1,48 @@
+        else:
             return self._minpad
 
+    def get_loc(self, axis):
```
See commit c4cfba7 where I adapted the annotation

athob commented on astropy/visualization/wcsaxes/frame.py:
```diff
@@ -110,25 +110,49 @@ def _update_normal(self):
         pixel = self._get_pixel()
+        normal_angle = self.normal_angle
+        # Flip pixels if element vectors are not well oriented
+        if np.all(np.abs((self.normal_angle - 135) % 360 - 180) <= 90.0, axis=0):
```
I've included that suggestion in batch commit 1dc8e12, but please let me know your thoughts regarding the explanation above

athob replied:
Quick update regarding this topic, I thought it might be beneficial to clarify what I described above in the source code documentation and comments, I did so in my latest commit 39d5d94. Let me know if that's okay, and if that helps better understand what's going on that line of code

Note that I had made a couple mistake in [my extended description 2 message above](https://github.com/astropy/astropy/pull/18749#discussion_r2479338484):
- I was saying the flipping happened if the `normal_angle` angles were all between -45 and 135 module 360, but the range I check is actually -135 and +45 modulo 360. In practice, that's what the expression `np.abs((normal_angle - 135) % 360 - 180) <= 90.0` achieves (considering -135=-315+180, it subtracts 315 degree - equivalent to -45 degree modulo 360 - then adds 180 to apply the modulo 360 before re-subtracting 180, and then verifying if the resulting quantity falls between -90 and +90).
- Also those `normal_angle` angles don't represent an "angle of the vectors that are normal to these spine vectors", they're just the angles between the spine vectors and the vertical vector oriented top to bottom (which I also explicitly documented in the `_update_normal` method).

Sorry about that confusion

athob committed:
Improving documentation
Clarifying the spine orientation criteria for `_barycentric_x_y_angle`

athob commented:
Hi @neutrinoceros, thanks for getting back to this! It's been a while I haven't checked it either to be fair so no worries. I'm doing a rebase now, and I'm gonna look into the squashing

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -396,3 +398,77 @@ def test_set_ticks_values():
+    ax.coords[0].set_axislabel("X", loc="right")
+    ax.coords[1].set_axislabel("Y", loc="top")
+    canvas.draw()
+    return fig
```
```suggestion
@figure_test
@pytest.mark.parametrize("xloc, yloc", [("left", "bottom"), ("right", "top")])
def test_left_bottom_loc_labels(xloc, yloc):
    fig = Figure(figsize=(6, 6))
    canvas = FigureCanvasAgg(fig)
    ax = WCSAxes(fig, [0.1, 0.1, 0.8, 0.8], aspect="equal")
    fig.add_axes(ax)
    ax.coords[0].set_axislabel_position("tb")
    ax.coords[1].set_axislabel_position("lr")
    ax.coords[0].set_axislabel("X", loc=xloc)
    ax.coords[1].set_axislabel("Y", loc=yloc)
    canvas.draw()
    return fig
```

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -396,3 +398,77 @@ def test_set_ticks_values():
+            "l": "bottom",
+            "r": "top",
+        },
+    )
```
I don't normally comment on style, but this really feels like overkill
```suggestion
    ax.coords[0].set_axislabel("X", loc={"t": "left", "b": "right"})
    ax.coords[1].set_axislabel("Y", loc={"l": "bottom", "r": "top"})
```

neutrinoceros reviewed (changes requested):
I think the next thing we need to do is getting your new image tests to pass, so let's get them ready first.

athob committed:
Apply suggestions from code review on figure tests

Co-authored-by: Clément Robert <cr52@protonmail.com>

neutrinoceros commented:
awesome. Output correctness is now trivial to assess so let's move forward with commiting their hashes. We have documentation on this process but it looks misleading to me, so I'll handle that step (and fix our docs).

neutrinoceros committed:
TST: save hashes for new image outputs

astrofrog commented on astropy/visualization/wcsaxes/frame.py:
```diff
@@ -105,30 +105,71 @@ def world(self, value):
-        # Find segment along which the mid-point lies
-        imin = np.searchsorted(d, d[-1] / 2.0) - 1
+        # Find segment along which the barycentric point lies
+        imin = min(0, np.searchsorted(d, dbary) - 1)
```
Should this be max rather than min?

astrofrog commented on astropy/visualization/wcsaxes/coordinate_helpers.py:
```diff
@@ -752,6 +768,24 @@ def set_axislabel(self, text, minpad=1, **kwargs):
+        if ignored := set(kwargs).intersection(protected_kw):
+            warnings.warn(
+                "The following low level keyword arguments "
+                "will be overwritten during the rendering."
```
```suggestion
                "will be overwritten during the rendering: "
```

(colon and add space)

astrofrog reviewed (changes requested):
Sorry for dropping the ball on this - I think this would be good to get over the finish line, so I've gone and done a detailed review. The two most important issues mentioned below are: (a) we should not use matplotlib private API, and (b) there is a potential min -> max bug that can cause issues. I will see if I can come up with an example that triggers the bug.

astrofrog commented:
Here's a minimal example that reproduces the min -> max bug. Maybe worth including a figure test like this to make sure that part of the code is properly exercised:

```python
import numpy as np
import matplotlib.pyplot as plt
from astropy.visualization.wcsaxes import WCSAxes
from astropy.visualization.wcsaxes.frame import BaseFrame


class CustomFrame(BaseFrame):
    """Frame with a single curved (arc) spine."""

    spine_names = 'b'
    _spine_auto_position_order = 'b'

    def update_spines(self):
        xmin, xmax = self.parent_axes.get_xlim()
        ymin, ymax = self.parent_axes.get_ylim()

        # Semicircular arc
        theta = np.linspace(0, np.pi, 200)
        x = (xmin + xmax) / 2 + (xmax - xmin) / 2 * np.cos(theta)
        y = ymin + (xmax - xmin) / 2 * np.sin(theta)
        self['b'].data = np.column_stack([x, y])

        super().update_spines()


fig = plt.figure(figsize=(15, 5))

for i, loc in enumerate(['left', 'center', 'right']):
    ax = WCSAxes(fig, [0.02 + i * 0.33, 0.1, 0.3, 0.8],
                 aspect='equal', frame_class=CustomFrame)
    fig.add_axes(ax)

    ax.coords[0].set_axislabel_visibility_rule('always')
    ax.coords[0].set_axislabel_position('b')
    ax.coords[0].set_axislabel('Label', loc=loc, color='red', size=14, weight='bold')

plt.savefig('bug_mwe.png', dpi=150, bbox_inches='tight', facecolor='white')
```

With the current ``min(``:

<img width="2208" height="718" alt="buggy" src="https://github.com/user-attachments/assets/d9d8b208-b95f-43f9-b320-db92714f92c3" />


Changing from ``min(`` to ``max(``:

<img width="2208" height="718" alt="bug_mwe" src="https://github.com/user-attachments/assets/a1a45831-d20c-41ee-8726-55130157b016" />

neutrinoceros commented:
@athob Feel free to re-open this whenever you'd like to come back to it.
